[AppConfiguration] Filtering key-values by tags: API Version 2023-11-01

avanigupta opened the pull request:
# Data Plane API - Pull Request
AppConfiguration Get KeyValues API needs to support a new query parameter for filtering by tags:​

### GET /kv?**tags={tagFilter1}&tags={tagFilter2}…&tags={tagFilter5}​**&api-version={}​

Where each `{tagFilter}` is an expression of the form `{tagName=tagValue}`. This expression string is opaque to the SDK.​

## Notes
- Multiple tag filters can be provided in a single request, result will be the AND of all those tag filters. ​
- Up to 5 tag filters are allowed. ​
- No changes expected in the response returned from server.​

## Examples

### GET /kv?tags=default=true&api-version=2023-11-01​

This will match all key-values with a tag named 'default' and value 'true'.​

### GET /kv?key=App1*&label=Prod&tags=group=g1&tags=default=true&api-version=2023-11-01​

This will match all key-values where:​
Keys start with App1​
Label = Prod​
Key-value has 2 tags named 'group' and 'default' and their values are 'g1' and 'true' respectively.​


## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/25816

Is this review for (select one):

- [ ] a private preview
- [] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/25816

avanigupta committed:
Previous Stable API

avanigupta committed:
Add tag filters to GET kv and update API version

JeffreyRichter commented:
I see that tags is just defines as a free-form JSON object. This will be exceedingly difficult to use in SDKs as there will be no code completion/type-safety. Also, I see a tagFilters query parameter that is also free-form JSON - this will require encoding the JSON string in some way making this incredibly difficult for customers to use. The usability of this is bad. I'm guessing that tags are string/string key/value pairs, right? Perhaps this can be expressed better.

avanigupta commented:
@JeffreyRichter thanks for your review. Yes, tags are string/string key-value pairs. We just need a way to let users filter by the exact tag name and value, and a way to accept up to 5 of such tag names and values. 
It looks like there's no way in openapi v2 to specify something like the query URL I originally mentioned in PR description. 

However, there is a way to get something like this with openapi v3: 
`GET /kv?tags[group]=g1&tags[default]=true​`
(Square brackets will be URL encoded, but leaving that out for readability here)

This would be the tag parameter definition for specifying the above:

```json
{
    "name": "tags",
    "in": "query",
    "description": "A filter used to match tags",
    "schema": {
      "type": "object"
    },
    "style": "deepObject",
    "explode": true
}
```

Is this a good approach?

avanigupta commented:
> Some service might offer a 'filter' query parameter with a string value like "group eq g1 and default eq true"

@JeffreyRichter Yes, I'm only aware of this pattern for filtering by ARM tags. The problem is that these are OData query parameters, which we are trying to avoid since our filtering logic is not as complicated. I've asked to see if there are other patterns that might be more suitable for our use case: https://stackoverflow.microsoft.com/questions/371499 

cc @drago-draganov

avanigupta commented:
@JeffreyRichter I reached out to ARM team and they only support OData queries for filtering. We want to avoid OData protocol because we already support many query parameters which enable filtering without OData queries. It would be odd to add OData support for just one query parameter. There isn't much guidance on this type of filtering in [Azure API guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#query-options) too.

Would it be possible to add support for OpenAPI v3 format? There's at least one service using openapi v3 already: [azure-rest-api-specs/blob/63c36a5f9b1fd55f233fb405c13480699c03f2f9/specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2023-05-15/inference.json](https://github.com/Azure/azure-rest-api-specs/blob/63c36a5f9b1fd55f233fb405c13480699c03f2f9/specification/cognitiveservices/data-plane/AzureOpenAI/inference/stable/2023-05-15/inference.json#L2)

avanigupta commented:
@JeffreyRichter it looks like Blob storage allows filtering on tags without using OData queries: [Find Blobs by Tags](https://learn.microsoft.com/en-us/rest/api/storageservices/find-blobs-by-tags-container?tabs=azure-ad#remarks)

Their API spec is quite simple:
https://github.com/Azure/azure-rest-api-specs/blob/6a2e3c7617314fe4ea7e5706da5437214e8a602b/specification/storage/data-plane/Microsoft.BlobStorage/preview/2021-12-02/blob.json#L11713-L11719

We'd like to follow a similar pattern, but without the dependency on ANSI SQL WHERE clause grammar. Instead of defining all tag filters in one `where` query param, we want to allow multiple `tags` params, where each `tags` param accepts one `name=value` filter. For example:

 `GET /kv?tags=group=g1&tags=default=true​`

To generalize:

### GET /kv?tags=tagName1=tagValue1&tags=tagName2=tagValue2&...&tags=tagName5=tagValue5

The API spec in OpenAPI V2 format will be:

```json
{
  "name": "tags",
  "in": "query",
  "description": "A filter used to match tags",
  "type": "array",
  "items": {
     "type": "string"
  },
  "collectionFormat": "multi",
  "maxItems": 5,
  "uniqueItems": true
}
```

Does that sound reasonable?

JeffreyRichter commented:
I'm on vacation this week so not fully focused. Technically this does work. We'd expect customers to actually do this: tags=tag1=val1,tag2=val2,tag3=val3. So 1 query param with comma-separated t=v pairs. I'd like to bounce this off the API stewarsdhip board and, if we all like it, we would formally add this to Azure's HTTP guidelines. I can bring it up next week when I'm back.

avanigupta commented:
@JeffreyRichter The issue with comma separated filters is that we already support this format for logical OR based filtering. But here we need logical AND filtering. For example, we already support key and label filters like this:

`GET /kv?key=abc,xyz` -> matches key name 'abc' OR 'xyz'
`GET /kv?label=dev,prod,test` -> matches label 'dev' OR 'prod' OR 'test'

JeffreyRichter commented:
As per the HTTP spec, having "qp=val1&qp=val2" is identical to having "qp=val1,val2".
Customers can use either syntax to get identical results.

So, I don't know how to you provide both AND and OR semantics to customers without a more complicated filter. Maybe tags=label=dev|prod|test,key=abc,xyz
So, label is dev OR prod or test AND key is abc OR xyz

Would this work for you?




  *   Jeffrey Richter<https://www.linkedin.com/in/jeffrichter/> (see Architecting Distributed Cloud Apps<http://aka.ms/RichterCloudApps> & Designing HTTP APIs<https://www.youtube.com/watch?v=9Ng00IlBCtw&list=PL9XzOCngAkqs4m0XdULJu_78nM3Ok3Q65&index=1>)


Azure API Stewardship: Guidelines<https://aka.ms/azapi/guidance>  |  Review process <https://aka.ms/azapi>  |  Recordings<https://aka.ms/azapi/teams>  |  Calendar<https://aka.ms/azapi/calendar>  |  Join the Interest Group<https://aka.ms/azapi/join-interest-group>

Azure SDK Architecture: Guidelines<https://azure.github.io/azure-sdk/general_introduction.html>  |  Review process <https://aka.ms/azsdk/archreview/process>  |  Recordings<https://aka.ms/azsdk/archreview/recordings>  |  Calendar<https://aka.ms/azsdk/archreview/calendar>  |  Join the Interest Group<https://aka.ms/azsdk/archreview/join-interest-group>

________________________________
From: Avani Gupta ***@***.***>
Sent: Thursday, September 28, 2023 11:24:30 AM
To: Azure/azure-rest-api-specs ***@***.***>
Cc: Jeffrey Richter ***@***.***>; Mention ***@***.***>
Subject: Re: [Azure/azure-rest-api-specs] [AppConfiguration] Filtering key-values by tags: API Version 2023-11-01 (PR #25861)


@JeffreyRichter<https://github.com/JeffreyRichter> The issue with comma separated filters is that we already support this format for logical OR based filtering. But here we need logical AND filtering. For example, we already support key and label filters like this:

GET /kv?key=abc,xyz -> matches key name 'abc' OR 'xyz'
GET /kv?label=dev,prod,test -> matches label 'dev' OR 'prod' OR 'test'

—
Reply to this email directly, view it on GitHub<https://github.com/Azure/azure-rest-api-specs/pull/25861#issuecomment-1739810544>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AARLJP4VGF6XE6X4DIG6J53X4W6F5ANCNFSM6AAAAAA46XFIHM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

avanigupta commented:
> tags=label=dev|prod|test,key=abc,xyz

This is still a contradicting semantic to our existing query parameters. In this approach, the comma is acting as logical AND, but our existing query parameters treat comma separated values as logical OR.



>  "qp=val1&qp=val2" is identical to having "qp=val1,val2"

That's not how our service behaves currently. That's because our qp is not an array, its a string. This is our swagger for such query parameters:
https://github.com/Azure/azure-rest-api-specs/blob/15ba5b8a954afb74a6bb8902038835f0cddb9930/specification/appconfiguration/data-plane/Microsoft.AppConfiguration/stable/2023-10-01/appconfiguration.json#L147-L157

JeffreyRichter commented:
I'm afraid you have a very big problem now as your service's semantics violate industry standard URL rules. For example, a customer using the Go programming language can't use your service correctly today as Go's standard URL package collapses all query parameter values for the same key together into just 1 key with an array of values (as is consistent with the URL rules). I wrote some Go code to prove this: https://go.dev/play/p/3Kwdfqwmpbu. My URL has 2 'tag' query parameters with different values, but Go's URL parser turns this into just 1 query parameter with an array with 2 values. You can easily see this if you Run the program. Some other programming languages MAY work but I don't know what other programming languages may not work. Also, some languages that do work today may stop working tomorrow should the language's URL type change how it handles this in the future which would be non-breaking as long as it maintains the URL rules. 

In addition, there is no good way to expose the same query parameter key multiple times in any SDK language. All SDKs would represent this as an array of values for a single key.

So, I think your team has to fix this as it is just broken & not SDK-usable, and some language clients will already not get the behavior they expect. Of course, fixing this is a breaking change but I don't see what choice your team has.

It's hard to solve your new problem without fixing the existing & bigger problem first.

drago-draganov commented:
> As per the HTTP spec, having "qp=val1&qp=val2" is identical to having "qp=val1,val2".

Do you mind referencing the said HTTP spec?

drago-draganov commented:
Not sure about de facto standard.

I don't think that Go example demonstrates identical parsing results:

tag=red&tag=blue -> array with 2 elements ["red", "blue"] - not CSV
tag=red,blue,green,yellow -> 1 string element "red,blue,green,yellow" - not an array with 4 elements
tag=red -> 1 string element "red"
tag=red,green&tag=blue -> array with 2 elements ["red,green", "blue"]


Re : Dotnet

HttpUtility.ParseQueryString is part of the legacy ASP.NET WebForms (System.Web). We carried it over for compat reasons alone.

Dotnet core and ASP.NET Core switched to WebUtilities.QueryHelpers, which diverges from System.Web logic.

QueryHelpers.ParseQuery produces Dictionary<string, StringValues>, where StringValues is a collection.

tag=red&tag=blue -> collection with 2 string elements ["red", "blue"]
tag=red,blue,green,yellow -> collection with 1 string element "red,blue,green,yellow"
tag=red -> collection with 1 string element "red"
tag=red,green,&tag=blue -> collection with 2 string elements ["red,green", "blue"]

The same applies to building query string APIs:

```cs
string uri = "http://contoso.com";

uri = QueryHelpers.AddQueryString(uri, "tag", "red");
uri = QueryHelpers.AddQueryString(uri, "tag", "blue");
uri = QueryHelpers.AddQueryString(uri, "tag", "yellow,green");
```

Result 
http://contoso.com?tag=red&tag=blue&tag=yellow,green

JeffreyRichter commented:
I see. Perhaps I was wrong all this time then. Thanks for setting me straight. It's good that this rarely comes up. So how do SDKs expose this then? I can't believe that we let customers specify the same query parameter multiple times. Do we make it an array with multiple string values?

JeffreyRichter commented:
OK, so based on my new understanding of how these query params work, what is the complete proposal for a customer to specify tags with AND & OR semantics? I'm looking for what some short documentation would look like to the HTTP/REST developer not using any SDK and what a fully fleshed out URL example would be showing AND & OR semantics.

avanigupta committed:
Add links for filter syntax reference

avanigupta committed:
Prettier whitespace change

avanigupta committed:
Move tags query param to end of list

jimmyca15 commented on specification/appconfiguration/data-plane/Microsoft.AppConfiguration/stable/2023-11-01/appconfiguration.json:
```diff
@@ -259,19 +259,19 @@
             "name": "tags",
             "in": "query",
-            "description": "A filter used to query by tags",
+            "description": "A filter used to query by tags. Syntax reference: https://learn.microsoft.com/en-us/azure/azure-app-configuration/rest-api-key-value#filtering",
```
Microsoft docs links shouldn't have the language. "en-us" here. 

Also, it would be a good idea to have an aka.ms link.

avanigupta replied:
Updated, thanks!

avanigupta committed:
Update with aka.ms links

avanigupta commented:
Hi @JeffreyRichter, the PR is ready for review. Could you please take another look and help to merge it?

jimmyca15 reviewed (changes requested):
we need tag filtering on snapshot creation.

avanigupta committed:
Tag filters for snapshots and revisions endpoint

avanigupta commented:
@JeffreyRichter, there are 2 scenarios which we missed during the review due to an oversight:
1. Our `revisions` endpoint is also designed to read a list of key-values similar to our `kv` endpoint. The idea is that if users can filter their key-values based on tags, they should be able to apply the same filters and get revisions of these key-values. I just added tag filters to the revisions endpoint as well. It looks exactly like the kv endpoint now:
#### GET /revisions?tags={tagFilter1}&tags={tagFilter2}…&tags={tagFilter5}​&api-version={}​

2. While creating snapshots, we support filtering key-values by their key and label. Since we are now adding support for filtering key-values by tags, the snapshot create operation needs to be updated too. Snapshot creation filters are provided in the request body, and I just added tag filters to the same.

Apologies for missing these earlier. Could you please review these additions in my last commit?

jimmyca15 commented:
@avanigupta we may need another review

JeffreyRichter commented:
I just reviewed it now and it all looks fine to me.

avanigupta commented:
Thank you @JeffreyRichter. In that case, can we merge the PR now?

JeffreyRichter commented:
The assignee (@anuchandy) can merge.

anuchandy commented:
@jimmyca15, could you approve since you _requested changes_ (needed by the CI)?

jimmyca15 commented:
@anuchandy approved
